test: unit tests for the back-end setup

s1m0n1au opened the pull request:
## Summary
- Set up a unit test framework/library for the back recommended by Cloudflare

## References
- https://developers.cloudflare.com/workers/testing/vitest-integration/write-your-first-test/
- https://developers.cloudflare.com/workers/testing/vitest-integration/test-apis/
- https://developers.cloudflare.com/workers/testing/vitest-integration/recipes/

s1m0n1au committed:
test: configure `vitest-pool-workers` with simple auth test

s1m0n1au committed:
fix: `vitest watch` fails on the second run

references:
- https://github.com/cloudflare/workers-sdk/issues/9913
- https://github.com/cloudflare/workers-sdk/issues/10506

s1m0n1au committed:
test: add unit test for a new profile and sent an email for confirmation

s1m0n1au committed:
test: add unit test for sign-up email exist

s1m0n1au committed:
test: add unit test for sign-up weak password

s1m0n1au committed:
test: add unit test for sign-up no name

s1m0n1au committed:
test: add unit test for sign-up incorrect token

s1m0n1au committed:
test: add unit test for Account activated successfully

s1m0n1au committed:
test: add unit test for login success

s1m0n1au committed:
test: add unit test for incorrect email/password

s1m0n1au committed:
test: add unit test for sign-out

s1m0n1au committed:
test: config d1 & add unit test for get existing user

references:
- https://developers.cloudflare.com/workers/testing/vitest-integration/test-apis/

s1m0n1au committed:
test: update unit test to use d1

s1m0n1au committed:
refactor: constant for api endpoint

s1m0n1au committed:
fix: lint

s1m0n1au committed:
fix: typo

s1m0n1au committed:
test: add the auth test with db

s1m0n1au committed:
fix: lint

s1m0n1au committed:
chore: update seed-data script for `vitest-pool-workers`

s1m0n1au committed:
revert: revert "chore: update seed-data script for `vitest-pool-workers`"

This reverts commit 63204bc33fccab6ac7842ac0be2930f8837f3af8.

s1m0n1au committed:
chore: update each SQL statement to single line for `vitest-pool-workers`

s1m0n1au committed:
fix: lint

s1m0n1au committed:
fix: lint

s1m0n1au committed:
refactor: better string manipulation

madcampos commented on vitest.config.ts:
```diff
@@ -1,11 +1,28 @@
+	const migrations = await readD1Migrations(migrationsPath);
+
+	const rawSeedSql = fs.readFileSync(path.join(process.cwd(), 'db/seeds/seed-data.sql'), 'utf8');
+	const seedSql = rawSeedSql.split('\n').filter((line) => !line.trim().startsWith('--')).join('');
```
<!-- Nitpicks are small, trivial, but necessary changes. Distinguishing nitpick comments significantly helps direct the reader's attention to comments requiring more involvement. -->
**nitpick:** This can be moved to a file on the seeds folder so that it does not clutter the configuration.

madcampos commented on package.json:
```diff
@@ -53,7 +44,8 @@
 		"@cloudflare/vite-plugin": "1.9.6",
-		"@cloudflare/vitest-pool-workers": "0.8.55",
+		"@cloudflare/vitest-pool-workers": "0.8.49",
+		"@cloudflare/workers-types": "^4.20251011.0",
```
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution. -->
**question:** Isn't this module basically deprecated and superseeded by the `wrangler types` command?

madcampos commented on package.json:
```diff
@@ -7,34 +7,25 @@
 	"scripts": {
 		"dev": "vite dev",
 		"prepare": "husky",
-
 		"start": "npm run dev",
-
 		"dev:cron": "wrangler dev --test-scheduled",
-
 		"generate:types": "wrangler types api/src/env.d.ts",
 		"db:setup": "wrangler d1 migrations apply vms-local --local",
 		"db:seed": "wrangler d1 execute vms-local --local --file=./db/seeds/seed-data.sql",
-
 		"build": "vite build",
 		"prepreview": "npm run build",
 		"preview": "vite preview",
-
 		"predeploy:dev": "npm run build",
 		"deploy:dev": "wrangler deploy --minify --env dev",
-
 		"predeploy:prod": "npm run build",
 		"deploy:prod": "wrangler deploy --minify --env production",
-
 		"format": "dprint fmt",
 		"format:check": "dprint check",
-
 		"typecheck:api": "tsc -p ./api --noEmit",
 		"typecheck:client": "tsc -p ./src --noEmit",
 		"typecheck": "npm run typecheck:api && npm run typecheck:client",
 		"lint:js": "eslint",
 		"lint": "npm run typecheck && npm run lint:js",
-
 		"test:coverage": "vitest run --coverage",
 		"test:unit": "vitest run",
 		"test:api-file": "hurl --continue-on-error --jobs 4 --max-time 300 --test --variables-file .dev.vars --variables-file .api-test.vars",
```
<!-- Nitpicks are small, trivial, but necessary changes. Distinguishing nitpick comments significantly helps direct the reader's attention to comments requiring more involvement. -->
**nitpick:** Please keep the line breaks for readability.

madcampos commented on api/tsconfig.json:
```diff
@@ -1,11 +1,18 @@
-		"lib": ["ESNext"],
+		"lib": [
+			"ESNext"
+		],
```
<!-- Nitpicks are small, trivial, but necessary changes. Distinguishing nitpick comments significantly helps direct the reader's attention to comments requiring more involvement. -->
**nitpick:** Please keep the formatting as it was before. The project formatter, `dprint`, should not change this as it was already formatted with the tool.

If you are using `prettier`, `biome`, or any other tool that is not `dprint`, please disable or uninstall those as they will mess up with the formatting already set by the project.

madcampos commented on api/src/routes/auth/data.test.ts:
```diff
@@ -1,139 +1,94 @@
+import { beforeAll, describe, expect, it } from 'vitest';
+import { activateProfile, checkActiveEmail, checkExistingEmail, getHeartbeatInfo, getLoginInfo } from './data.ts';
+
+const USER = {
```
<!-- Issues highlight specific problems with the subject under review. These problems can be user-facing or behind the scenes. It is strongly recommended to pair this comment with a suggestion. If you are not sure if a problem exists or not, consider leaving a question. -->
**issue:** Use a more descriptive variable name. `USER` in this context doesn't give any clue to what this variable will be used for. Is it a single user used for **ALL** tests? Are there diferent users with diferent data?

Also, keep the data related to a test closer to the test using that data. It improves locality and makes tests more readable and easier to understand.

Tests are the type of code that _should_ be written in a verbose way to be very descriptive and avoid bugs by omitting or obscuring code. Make it simple, make it straight forward, and make it clear to someone reading the test what it checks for.

madcampos commented on api/src/routes/auth/data.test.ts:
```diff
@@ -1,139 +1,94 @@
-
-	beforeEach(() => {
-		env = new MockEnvBindings();
+describe('Auth data functions', () => {
```
<!-- Suggestions propose improvements to the current subject. It's important to be explicit and clear on what is being suggested and why it is an improvement. Consider using patches and the blocking or non-blocking decorations to further communicate your intent. -->
**suggestion:** while refactoring tests, add more descriptive descriptions to the tests.

Imagine someone is just running the tests without reading the files. The output should be descriptive to understand what is going on and what is being tested.

madcampos commented on api/src/routes/auth/data.test.ts:
```diff
@@ -1,139 +1,94 @@
+describe('Auth data functions', () => {
+	beforeAll(async () => {
+		await applyD1Migrations(env.Database, env.TEST_MIGRATIONS);
+		await env.Database.exec(env.SEED_SQL);
```
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution. -->
**question:** Does this apply to every single test? If so, shouldn't it be called in the outer scope instead of here?

Also, what executing this SQL does? Does it overwrite existing rows? Does it seeds again a new database? Does it `DROP` the existing rows and then re-`INSERT` them?

madcampos commented on api/src/routes/auth/data.test.ts:
```diff
@@ -1,139 +1,94 @@
-			});
-
-			const result = await getLoginInfo(env.Database, sampleUser.email);
+		it('returns credentials for an active, non-deleted profile', async () => {
```
<!-- Suggestions propose improvements to the current subject. It's important to be explicit and clear on what is being suggested and why it is an improvement. Consider using patches and the blocking or non-blocking decorations to further communicate your intent. -->
**suggestion:** Again: describe what are you trying to test, not the implementation of the test.

madcampos commented on api/src/routes/auth/data.test.ts:
```diff
@@ -1,139 +1,94 @@
-				access: sampleUser.access
+				id: USER.id,
+				password: USER.password,
+				access: USER.access
```
<!-- Issues highlight specific problems with the subject under review. These problems can be user-facing or behind the scenes. It is strongly recommended to pair this comment with a suggestion. If you are not sure if a problem exists or not, consider leaving a question. -->
**issue:** Keep the user information _in the test_, and not in a variable referenced elsewhere.

madcampos commented on api/src/routes/auth/data.test.ts:
```diff
@@ -1,139 +1,94 @@
 		});
 
-		test('returns null if no matching user found', async () => {
+		it('returns null if no matching user found', async () => {
```
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution. -->
**question:** Why the change from `test` to `it`?

madcampos commented on api/src/routes/auth/data.test.ts:
```diff
@@ -1,139 +1,94 @@
 		});
 
-		test('returns null if user is not found', async () => {
+		it('returns null if user is not found', async () => {
```
Same here as the others: locality of the test data and change from `test` to `it`

madcampos commented on api/src/routes/auth/data.test.ts:
```diff
@@ -1,139 +1,94 @@
 		});
 	});
 
 	describe('checkActiveEmail', () => {
-		test('returns true if email is active', async () => {
-			env.setDatabase({
-				first: <T>() => ({ email: sampleUser.email }) as T
-			});
-
-			const result = await checkActiveEmail(env.Database, sampleUser.email);
+		it('returns true if email is active', async () => {
+			const result = await checkActiveEmail(env.Database, USER.email);
 
 			expect(result).toBe(true);
 		});
 
-		test('returns false if email is inactive', async () => {
-			env.setDatabase({
-				first: <T>() => null as T
-			});
-
+		it('returns false if email is inactive', async () => {
 			const result = await checkActiveEmail(env.Database, 'inactive@example.com');
 
 			expect(result).toBe(false);
 		});
 	});
 
 	describe('activateProfile', () => {
-		test.todo('returns true when update succeeds');
-		test.todo('returns false when update fails');
+		it('returns true when update succeeds', async () => {
+			const result = await activateProfile(env.Database, USER.email);
+
+			expect(result).toBe(true);
+		});
+
+		it.todo('returns false when update fail');
 	});
 });
```
Same here: why the change from `test` to `it`?

madcampos commented on api/src/routes/auth/index.db.test.ts:
```diff
@@ -0,0 +1,142 @@
+import { applyD1Migrations, env } from 'cloudflare:test';
+import { beforeAll, describe, expect, it, vi } from 'vitest';
+
+import type * as EmailModule from '../../email/index.ts';
```
<!-- Nitpicks are small, trivial, but necessary changes. Distinguishing nitpick comments significantly helps direct the reader's attention to comments requiring more involvement. -->
**nitpick:** Please use explicit imports. When you import or export into a property bag it is not immediatly evident what you are importing, what you are using and how it is used.

It makes harder to clean-up unused code and adds indirection to the the code.

If you have a name clash use this construct instead:
```typescript
import { nameClash as notNameClash } from './module.ts';
```

madcampos commented on api/src/routes/auth/index.db.test.ts:
```diff
@@ -0,0 +1,142 @@
+		...actual,
+		sendAccountConfirmationEmail: vi.fn().mockResolvedValue(undefined)
+	};
+});
```
<!-- Issues highlight specific problems with the subject under review. These problems can be user-facing or behind the scenes. It is strongly recommended to pair this comment with a suggestion. If you are not sure if a problem exists or not, consider leaving a question. -->
**issue:** Don't.

Mocked modules create issues because they fake what the module does, they often deviate from the implementation causing issues where the test pass but the actual implementation has errors.

If you need control over the result of a function call while still keeping it "black boxed", use dependency injection to provide a callback to be called after the execution with the resulting value.

Furthermore, for testing emails, by default, no email will be sent and the email contents will be output to the console. You can control the behavior of email sending through environment variables.

madcampos commented on api/src/routes/auth/index.db.test.ts:
```diff
@@ -0,0 +1,142 @@
+	};
+});
+
+const loadApp = async () => (await import('../../index.ts')).app;
```
<!-- Suggestions propose improvements to the current subject. It's important to be explicit and clear on what is being suggested and why it is an improvement. Consider using patches and the blocking or non-blocking decorations to further communicate your intent. -->
**suggestion:** Prefer a regular import over dynamic import.

Also: use a regular function declaration instead of an arrow function.

Furthermore: if the app code itself (and it's internal state) doesn't change between tests, keep a global refernece instead of recrating the app every time, this is just wasteful and not needed.

madcampos commented on api/src/routes/auth/index.db.test.ts:
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution. -->
**question:** Why is the file named like this? Shouldn't it be just `index.test.ts`?

madcampos commented on api/src/routes/auth/index.db.test.ts:
```diff
@@ -0,0 +1,142 @@
+	beforeAll(async () => {
+		await applyD1Migrations(env.Database, env.TEST_MIGRATIONS);
+		await env.Database.exec(env.SEED_SQL);
+	});
```
Same question as the other test file: if it is applied to all tests, shouldn't this be called in the outer scope?

madcampos commented on api/src/routes/auth/index.db.test.ts:
```diff
@@ -0,0 +1,142 @@
+	SIGN_IN: '/api/auth/sign-in',
+	SIGN_OUT: '/api/auth/sign-out',
+	ACTIVATE: '/api/auth/activate'
+} as const;
```
<!-- Suggestions propose improvements to the current subject. It's important to be explicit and clear on what is being suggested and why it is an improvement. Consider using patches and the blocking or non-blocking decorations to further communicate your intent. -->
**suggestion:** Here is a place where extracting information to a variable does not help and can actually hurt DX.

You can hard code the endpoints to the tests, as changing the endpoint breaks the tests and that is expected behavior.

When the person runs the test the information will be there already improving locality of the code instead of hidden away in a reference that needs to be navigated to, where it could very much be right by the test itself.

madcampos commented on api/src/routes/auth/index.db.test.ts:
```diff
@@ -0,0 +1,142 @@
+			const [firstActivationTokenKey] = tokens.keys;
+			if (!firstActivationTokenKey) {
+				throw new Error('Activation token was not created');
+			}
```
<!-- Issues highlight specific problems with the subject under review. These problems can be user-facing or behind the scenes. It is strongly recommended to pair this comment with a suggestion. If you are not sure if a problem exists or not, consider leaving a question. -->
**issue:** This error throwing here seems odd... Tests should assert something instead.

Also, the KV stores should be flushed between every test execution to avoid mising up data between tests.

Every test should be as self contained as possible. Meaning: it should execute all the steps required for an action from an inital state.

madcampos commented on api/src/routes/auth/index.db.test.ts:
```diff
@@ -0,0 +1,142 @@
+				expect.anything(),
+				expect.objectContaining({
+					email: 'new@new.new',
+					// apiKey: env.RESEND_API_KEY,
```
<!-- Nitpicks are small, trivial, but necessary changes. Distinguishing nitpick comments significantly helps direct the reader's attention to comments requiring more involvement. -->
**nitpick:** Remove unused code to tidy up the codebase. If you need to go back to it, the git history is here to help you.

madcampos commented on api/src/routes/auth/index.db.test.ts:
```diff
@@ -0,0 +1,142 @@
+		});
+
+		describe('POST /api/auth/sign-in', () => {
+			it('returns 201 and creates a session when credentials are valid', async () => {
```
<!-- Nitpicks are small, trivial, but necessary changes. Distinguishing nitpick comments significantly helps direct the reader's attention to comments requiring more involvement. -->
**nitpick:** Better test description

madcampos commented on api/src/routes/auth/index.db.test.ts:
```diff
@@ -0,0 +1,142 @@
+				}, env);
+
+				expect(response.status).toBe(StatusCodes.CREATED);
+				await expect(response.json()).resolves.toEqual({ message: 'Sign in successful' });
```
<!-- Issues highlight specific problems with the subject under review. These problems can be user-facing or behind the scenes. It is strongly recommended to pair this comment with a suggestion. If you are not sure if a problem exists or not, consider leaving a question. -->
**issue:** Do not validate messages as they may change. The status code should be enough of a test if the operation succeeded or not.

Messages should be informative to the consumer of the API but not depended upon as they are ephemeral and subject to change. All good API designs should provide a way for code to infer the result of an operation, and sometimes a message to help users understand what happened. In this case the API provides HTTP status codes to inform the result of an operation.

madcampos commented on api/src/routes/auth/index.db.test.ts:
```diff
@@ -0,0 +1,142 @@
+				);
+
+				expect(response.status).toBe(StatusCodes.UNAUTHORIZED);
+				await expect(response.json()).resolves.toEqual({ message: 'Either your email/password combination is invalid, or your account is not active' });
```
Same here: https://github.com/torontojs/communityhub/pull/284/files#r2429868980

madcampos commented on api/src/routes/auth/index.test.ts:
```diff
@@ -1,99 +1,398 @@
+import type * as PasswordStrengthModule from '../../utils/passwordStrengthCheck.ts';
+import { StatusCodes } from '../../utils/responses.ts';
+import type * as ProfileModule from '../profile/data.ts';
+import type * as AuthDataModule from './data.ts';
```
Same issue as the other imports: prefer explicit imports.

madcampos commented on api/src/routes/auth/index.test.ts:
```diff
@@ -1,99 +1,398 @@
+	SIGN_IN: '/api/auth/sign-in',
+	SIGN_OUT: '/api/auth/sign-out',
+	ACTIVATE: '/api/auth/activate'
+} as const;
```
Same as here: https://github.com/torontojs/communityhub/pull/284/files#r2429841323

madcampos commented on api/src/routes/auth/index.test.ts:
```diff
@@ -1,99 +1,398 @@
+	password: 'H0lyGr@il42!',
+	access: 'volunteer' as const,
+	status: 'active'
+} as const;
```
Same as here: https://github.com/torontojs/communityhub/pull/284/files#r2429771544

madcampos commented on api/src/routes/auth/index.test.ts:
```diff
@@ -1,99 +1,398 @@
+		revalidateSession: vi.fn().mockResolvedValue(null),
+		createSession: vi.fn().mockResolvedValue(undefined)
+	};
+});
```
Same as here: https://github.com/torontojs/communityhub/pull/284/files#r2429827129

madcampos commented on api/src/routes/auth/index.test.ts:
```diff
@@ -1,99 +1,398 @@
+});
 
-	test.todo('failure: activation should not be successful if token expired');
+const loadApp = async () => (await import('../../index.ts')).app;
```
Same as here: https://github.com/torontojs/communityhub/pull/284/files#r2429829979

madcampos commented on api/src/routes/auth/index.test.ts:
```diff
@@ -1,99 +1,398 @@
 
-		expect(json.message).toBe('Account activated successfully');
+		const sessionKeys = await env.SessionTokens.list();
+		await Promise.all(sessionKeys.keys.map(async ({ name }) => env.SessionTokens.delete(name)));
```
This should be done on the other tests that reference KV

madcampos commented on api/src/routes/auth/index.test.ts:
```diff
@@ -1,99 +1,398 @@
-		);
+describe('Authentication routes', () => {
+	beforeEach(async () => {
+		vi.clearAllMocks();
```
Again: don't use mocks.

madcampos commented on api/src/routes/auth/index.test.ts:
```diff
@@ -1,99 +1,398 @@
+				{
+					method: 'POST',
+					headers: { 'Content-Type': 'application/json' },
+					body: JSON.stringify(USER)
```
<!-- Questions are appropriate if you have a potential concern but are not quite sure if it's relevant or not. Asking the author for clarification or investigation can lead to a quick resolution. -->
**question:** Does `body` expect a string?

s1m0n1au replied:
Yes. The body should be JSON string Payload.

madcampos commented on api/src/routes/auth/index.test.ts:
```diff
@@ -1,99 +1,398 @@
+			expect(response.status).toBe(StatusCodes.OKAY);
+			await expect(response.json()).resolves.toEqual({
+				message: 'Created a new profile and sent an email for confirmation'
+			});
```
Same here: https://github.com/torontojs/communityhub/pull/284/files#r2429868980

madcampos commented on api/src/routes/auth/index.test.ts:
```diff
@@ -1,99 +1,398 @@
+				email: USER.email,
+				name: USER.name,
+				password: 'hashed-password'
+			});
```
<!-- Issues highlight specific problems with the subject under review. These problems can be user-facing or behind the scenes. It is strongly recommended to pair this comment with a suggestion. If you are not sure if a problem exists or not, consider leaving a question. -->
**issue:** Whenever possible, don't use dynamic imports.

They make code harder to debug and harder to test as they make the code less deterministic.

madcampos commented on api/src/routes/auth/index.test.ts:
```diff
@@ -1,99 +1,398 @@
+			const [firstActivationTokenKey] = tokens.keys;
+			if (!firstActivationTokenKey) {
+				throw new Error('Activation token was not created');
+			}
```
Same as here: https://github.com/torontojs/communityhub/pull/284/files#r2429853036

madcampos commented on api/src/routes/auth/index.test.ts:
```diff
@@ -1,99 +1,398 @@
+				id: '3c5123c0-8548-4a02-a83c-32e9ce67eae8'
+			});
+
+			const { sendAccountConfirmationEmail } = await import('../../email/index.ts');
```
Same as here: https://github.com/torontojs/communityhub/pull/284/files#r2429899595

madcampos commented on api/src/routes/auth/index.test.ts:
```diff
@@ -1,99 +1,398 @@
+		it('returns 200 but skips creation when email already exists', async () => {
+			const app = await loadApp();
+
+			const { checkExistingEmail } = await import('../../routes/auth/data.ts');
```
Same as here: https://github.com/torontojs/communityhub/pull/284/files#r2429899595

madcampos commented on api/src/routes/auth/index.test.ts:
```diff
@@ -1,99 +1,398 @@
+			const app = await loadApp();
+
+			const { checkExistingEmail } = await import('../../routes/auth/data.ts');
+			vi.mocked(checkExistingEmail).mockResolvedValueOnce(true);
```
**issue**: mocks

madcampos commented on api/src/routes/auth/index.test.ts:
```diff
@@ -1,99 +1,398 @@
+			expect(response.status).toBe(StatusCodes.OKAY);
+
+			const { insertProfile } = await import('../../routes/profile/data.ts');
+			const { sendAccountConfirmationEmail } = await import('../../email/index.ts');
```
Same as here: https://github.com/torontojs/communityhub/pull/284/files#r2429899595

madcampos commented on api/src/routes/auth/index.test.ts:
```diff
@@ -1,99 +1,398 @@
+		it('returns 422 when password fails strength check', async () => {
+			const app = await loadApp();
+
+			const { passwordStrengthCheck } = await import('../../utils/passwordStrengthCheck.ts');
```
Same as here: https://github.com/torontojs/communityhub/pull/284/files#r2429899595

madcampos commented on api/src/routes/auth/index.test.ts:
```diff
@@ -1,99 +1,398 @@
+			const keys = await env.ActivationTokens.list();
+			await Promise.all(keys.keys.map(async ({ name }) => env.ActivationTokens.delete(name)));
+
+			const invalidToken = '3c5123c0-8548-4a02-a83c-32e9ce67eae8';
```
<!-- Issues highlight specific problems with the subject under review. These problems can be user-facing or behind the scenes. It is strongly recommended to pair this comment with a suggestion. If you are not sure if a problem exists or not, consider leaving a question. -->
**issue:** Please use an UUID with all zeros as the reference for non existent UUID. It is a valid UUID in the sense that it follows the format, but should not be a valid value in the application. As a bonus it is a very explicit value.

madcampos commented on api/src/routes/auth/index.test.ts:
```diff
@@ -1,99 +1,398 @@
+			expect(response.status).toBe(StatusCodes.CREATED);
+			await expect(response.json()).resolves.toEqual({ message: 'Sign in successful' });
+
+			const { createSession } = await import('../../../src/utils/auth.ts');
```
Same as here: https://github.com/torontojs/communityhub/pull/284/files#r2429899595

madcampos commented on api/src/routes/auth/index.test.ts:
```diff
@@ -1,99 +1,398 @@
+			const { createSession } = await import('../../../src/utils/auth.ts');
+			expect(createSession).toHaveBeenCalledWith(
+				expect.objectContaining({
+					id: '3c5123c0-8548-4a02-a83c-32e9ce67eae8',
```
<!-- Issues highlight specific problems with the subject under review. These problems can be user-facing or behind the scenes. It is strongly recommended to pair this comment with a suggestion. If you are not sure if a problem exists or not, consider leaving a question. -->
**issue:** Do not depend on a value from another test. The tests can (and should) be run out of order and in parallel.

Each test should be independent and self contained.

madcampos commented on api/src/routes/auth/index.test.ts:
```diff
@@ -1,99 +1,398 @@
+			const app = await loadApp();
+
+			const { getLoginInfo } = await import('../../routes/auth/data.ts');
+			vi.mocked(getLoginInfo).mockResolvedValueOnce(null);
```
Same as here: https://github.com/torontojs/communityhub/pull/284/files#r2429899595

madcampos commented on api/src/routes/auth/index.test.ts:
```diff
@@ -1,99 +1,398 @@
+			);
+
+			expect(response.status).toBe(StatusCodes.UNAUTHORIZED);
+			await expect(response.json()).resolves.toEqual({ message: 'Either your email/password combination is invalid, or your account is not active' });
```
Same here: https://github.com/torontojs/communityhub/pull/284/files#r2429868980

madcampos commented on api/src/routes/auth/index.test.ts:
```diff
@@ -1,99 +1,398 @@
+		it('returns 401 login with incorrect password', async () => {
+			const app = await loadApp();
+
+			const { validatePassword } = await import('../../utils/password-hashing.ts');
```
Same as here: https://github.com/torontojs/communityhub/pull/284/files#r2429899595

madcampos commented on api/src/routes/auth/index.test.ts:
```diff
@@ -1,99 +1,398 @@
+			);
+
+			expect(response.status).toBe(StatusCodes.UNAUTHORIZED);
+			await expect(response.json()).resolves.toEqual({ message: 'Either your email/password combination is invalid, or your account is not active' });
```
Same here: https://github.com/torontojs/communityhub/pull/284/files#r2429868980

madcampos commented on api/src/routes/auth/index.test.ts:
```diff
@@ -1,99 +1,398 @@
+				originalExpiry: new Date().toISOString()
+			};
+
+			const { revalidateSession } = await import('../../utils/auth.ts');
```
Same as here: https://github.com/torontojs/communityhub/pull/284/files#r2429899595

madcampos commented on api/src/routes/auth/index.test.ts:
```diff
@@ -1,99 +1,398 @@
+			const { revalidateSession } = await import('../../utils/auth.ts');
+
+			await env.SessionTokens.put(token, JSON.stringify(sessionPayload));
+			vi.mocked(revalidateSession).mockResolvedValue(sessionPayload);
```
**isue**: mocks

madcampos commented on api/src/routes/auth/index.test.ts:
```diff
@@ -1,99 +1,398 @@
+
+		it('returns 401 when the session is missing or invalid', async () => {
+			const app = await loadApp();
+			const { revalidateSession } = await import('../../utils/auth.ts');
```
Same as here: https://github.com/torontojs/communityhub/pull/284/files#r2429899595

madcampos commented on api/src/routes/auth/index.test.ts:
```diff
@@ -1,99 +1,398 @@
+			const app = await loadApp();
+			const { revalidateSession } = await import('../../utils/auth.ts');
+
+			vi.mocked(revalidateSession).mockResolvedValue(undefined);
```
**issue**: mocks

madcampos reviewed (changes requested):
Please review the code, there are some serious anti-patterns going on.

A lot of the test code is following the same patterns, so I may have missed some, so take it that everywhere you see something that should be changed, change it.

s1m0n1au committed:
style: project prefer format

s1m0n1au committed:
refactor: cleaner `vitest.config.ts` setup

s1m0n1au committed:
fix: missing `RESEND_API_KEY` during testing

s1m0n1au committed:
test: remove mock test

s1m0n1au committed:
test: improve test clarity and isolation

- rename cases to produce self-explanatory CLI output
- co-locate fixtures with their tests for easier maintenance
- assert on return values rather than brittle messages
- reset the database before each run to avoid leaking state

s1m0n1au committed:
test: improve test clarity and isolation

- rename cases to produce self-explanatory CLI output
- co-locate fixtures with their tests for easier maintenance
- assert on return values rather than brittle messages
- reset the database before each run to avoid leaking state
- flush the KV stores to avoid mising up data between tests
- remove mock

s1m0n1au committed:
test: use nil UUID for non-existent references for clarity

s1m0n1au committed:
chore: remove deprecated package

madcampos commented on package.json:
```diff
@@ -72,7 +72,7 @@
 		"typescript": "5.8.3",
 		"vite": "^7.0.4",
-		"vitest": "3.2.4",
+		"vitest": "~3.2.0",
```
<!-- Issues highlight specific problems with the subject under review. These problems can be user-facing or behind the scenes. It is strongly recommended to pair this comment with a suggestion. If you are not sure if a problem exists or not, consider leaving a question. -->
**issue:** Please prefer `^` or fixed version numbers.

madcampos commented on package.json:
```diff
@@ -53,7 +53,7 @@
 	"devDependencies": {
 		"@cloudflare/vite-plugin": "1.9.6",
-		"@cloudflare/vitest-pool-workers": "0.8.55",
+		"@cloudflare/vitest-pool-workers": "0.8.49",
```
<!-- Issues highlight specific problems with the subject under review. These problems can be user-facing or behind the scenes. It is strongly recommended to pair this comment with a suggestion. If you are not sure if a problem exists or not, consider leaving a question. -->
**issue:** Unless it is strictly necessary, don't downgrade packages.

bcha92 replied:
**issue:** agreed. Downgrading packages are not recommended as it would introduce potentially a host of bugs and vulnerabilities that otherwise have been patched in later version. There are exceptions of course, but for minor (non-breaking) downgrades like these, it is generally not recommended.

Conversely, upgrading packages also warrants caution because we could be introducing breaking changes (i.e. `1.0.0` to `2.0.0`) that may render the current logic to fail or malfunction when executed. But rule of thumb is (but not always) to upgrade to the version just prior to the breaking changes (i.e. `1.0.4` to `1.1.6`)

bcha92 replied:
I see there's an annoying console warning log for an error when cleaning up temp directories while attempting to use `npm run test` on v0.8.55 that doesn't appear on v0.8.49. It's not a blocker per se for testing, but it's annoying having to scroll up a bit to see if the tests passed or failed.

After some google searching, I think this `console.warn` is only seen by Windows users and there is no clear fix. Then again this SDK is a bit experimental to say at the least with more weird stuff worse than this, so I think we can just revert the change here back to the original (latest) version and put in a non-urgent ticket for this down the road.

madcampos replied:
@bcha92 is there a report already for this warning? If there is, please link the issue so we can also keep track of that.

If not, would be a good idea to create a bug report. Please try to be as specific as possible with the report and provide as much info as you can.

bcha92 replied:
@madcampos only in-console (tried running with the `"@cloudflare/vitest-pool-workers": "0.8.55"` version reverted)
<img width="1509" height="356" alt="Screenshot 2025-10-27 010200" src="https://github.com/user-attachments/assets/967b2821-161a-4c3e-a688-922e684b8482" />

madcampos replied:
@bcha92 raise an issue on the cloudflare repo, this seems like a library issue.
Also, if you have the resources, would be good to try on linux or mac to see if it is a Windows only issue.

bcha92 replied:
hey @madcampos issue on the cloudflare repo is already up regarding failure to remove temp directory. Temporary solution is to downgrade the `"@cloudflare/vitest-pool-workers"` package to version `"0.8.49"`

https://github.com/cloudflare/workers-sdk/issues/10511

I will close this comment if that is okay.

madcampos commented on wrangler.toml:
<!-- Issues highlight specific problems with the subject under review. These problems can be user-facing or behind the scenes. It is strongly recommended to pair this comment with a suggestion. If you are not sure if a problem exists or not, consider leaving a question. -->
**issue:** The variables are set in [`.dev.vars.example`](https://github.com/torontojs/communityhub/blob/dev/.dev.vars.example) and should not be updated in the `wrangler.toml` file.

On your local environment you should copy the `.dev.vars.example` file to `.dev.vars` and adjust the values as needed.

See: [Setup Guide](https://github.com/torontojs/communityhub/blob/dev/docs/contributing-guide/setup.md?plain=1#L14-L17)

madcampos reviewed (changes requested):
I haven't finished reviewing the tests themselves, but here are a couple of notes.

s1m0n1au committed:
revert: "fix: missing `RESEND_API_KEY` during testing"

This reverts commit 6544a68cc21504c6276b6364532a403258adce99.

s1m0n1au committed:
chore: fixed version number

s1m0n1au committed:
fix: typo

bcha92 commented on api/src/routes/auth/data.test.ts:
```diff
@@ -1,139 +1,90 @@
-	getLoginInfo
-} from './data.ts';
-
-interface User {
-	id: string;
-	email: string;
-	password: string;
-	access: AccessLevel;
-	name: string;
-	avatar?: string;
```
**comment:** I think this is a good opportunity to move this interface into a schema where we can re-inforce this strict typing in both the testing environment and in the logic.
**issue:** I don't see the interface being re-introduced to re-inforce strict typing, would that cause some issues down the road when testing for values not consistent with the user attributes?

madcampos replied:
+1: This interface should be inferred from an schema already, and if there is a need for slightly altering the interface, that's what the `Omit`, `Pick`, and `Partial` [utility types](https://www.typescriptlang.org/docs/handbook/utility-types.html) are for.

madcampos replied:
One addition to that: as the parameters for the API are already covered by zod (or should be), adding tests to that is just bloat, as the library should already be tested.

We should only add tests to things that fall outside of library code, meaning, things we add _after_ the zod validation runs and is successful.

bcha92 reviewed (approved):
Looks good (just update base branch)

madcampos commented on api/src/routes/auth/data.test.ts:
```diff
@@ -1,139 +1,90 @@
+	await applyD1Migrations(env.Database, env.TEST_MIGRATIONS);
+});
 
+beforeEach(async () => {
+	await env.Database.exec(env.SEED_SQL);
+});
```
**question**: won't this cause errors?

Like, if you are seeding the database again before each test, you will try to insert the same ID over and over again, causing a constraint validation to fail.

Is there any "cleanup" process to remove the seeded data?

s1m0n1au replied:
According to the docs, when `isolatedStorage: true` default have the  "cleanup" process to remove the seeded data.
<img width="698" height="261" alt="image" src="https://github.com/user-attachments/assets/e613b522-16a7-497a-9fe2-6300de64abc0" />
<img width="452" height="292" alt="image" src="https://github.com/user-attachments/assets/89075a90-bfc6-4709-9f26-b24b86d8547b" />

Reference:
https://developers.cloudflare.com/workers/testing/vitest-integration/configuration/ 
https://developers.cloudflare.com/workers/testing/vitest-integration/isolation-and-concurrency/

madcampos reviewed (approved):
Approved with a non blocking question :)
